SharpMUSH.Implementation/Functions/CommunicationFunctions.cs

```diff
@@ -144,61 +145,198 @@ public static async ValueTask<CallState> NoSpoofLocationEmit(IMUSHCodeParser par
+	[SharpFunction(Name = "nsoemit", MinArgs = 2, MaxArgs = 2, Flags = FunctionFlags.Regular)]
 	public static ValueTask<CallState> NoSpoofOmitEmit(IMUSHCodeParser parser, SharpFunctionAttribute _2)
 	{
 		throw new NotImplementedException();
 	}
 
-	[SharpFunction(Name = "NSPEMIT", MinArgs = 2, MaxArgs = 2, Flags = FunctionFlags.Regular)]
-	public static ValueTask<CallState> NoSpoofPrivateEmit(IMUSHCodeParser parser, SharpFunctionAttribute _2)
+	[SharpFunction(Name = "nspemit", MinArgs = 2, MaxArgs = 2, Flags = FunctionFlags.Regular | FunctionFlags.HasSideFX)]
+	public static async ValueTask<CallState> NoSpoofPrivateEmit(IMUSHCodeParser parser, SharpFunctionAttribute _2)
 	{
-		throw new NotImplementedException();
+		if (!Configuration!.CurrentValue.Function.FunctionSideEffects)
+		{
+			return new CallState(Errors.ErrorNoSideFx);
+		}
+		
+		var executor = await parser.CurrentState.KnownExecutorObject(Mediator!);
+		var recipients = parser.CurrentState.Arguments["0"].Message!.ToPlainText();
+		var message = parser.CurrentState.Arguments["1"].Message!;
+		
+		// Determine notification type based on nospoof permissions
+		var notificationType = await PermissionService!.CanNoSpoof(executor)
+			? INotifyService.NotificationType.NSAnnounce
+			: INotifyService.NotificationType.Announce;
+		
+		// Check if first argument is an integer list (port list)
+		if (IsIntegerList(recipients))
+		{
+			// Handle port-based messaging
+			// IsIntegerList already validated these are parseable integers
+			var ports = recipients.Split(' ', StringSplitOptions.RemoveEmptyEntries)
+				.Select(long.Parse)
+				.ToArray();
+			
+			await NotifyService!.Notify(ports, message, executor, notificationType);
+		}
+		else
+		{
+			// Handle object/player-based messaging
+			var recipientList = ArgHelpers.NameList(recipients);
+			
+			foreach (var recipient in recipientList)
+			{
+				await recipient.Match(
+					async dbref =>
+					{
+						// Get the object to check permissions
+						var maybeTarget = await Mediator!.Send(new GetObjectNodeQuery(dbref));
+						if (!maybeTarget.TryPickT0(out var target, out var _))
+						{
+							return;
+						}
+						
+						// Check if executor can interact with the target
+						if (!await PermissionService!.CanInteract(target, executor, InteractType.Hear))
+						{
+							return;
+						}
+						
+						await NotifyService!.Notify(dbref, message, executor, notificationType);
+					},
+					async name =>
+					{
+						var maybeFound = await LocateService!.LocatePlayer(parser, executor, executor, name);
+						if (maybeFound.TryPickT0(out var player, out var _))
+						{
+							// Check if executor can interact with the player
+							if (!await PermissionService!.CanInteract(player, executor, InteractType.Hear))
+							{
+								return;
+							}
+							
+							await NotifyService!.Notify(player, message, executor, notificationType);
+						}
```

Comment:
Why are you `return`-ing early in a foreach when the remainder should still be handled?

Use `continue` instead.